uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-assignment.controller.js

```diff
@@ -259,6 +261,91 @@ define([
       });
     }
 
+    /**
+     * Returns the default assignee for the activity type provided.
+     *
+     * @param {Object} activityType the activity type definition.
+     * @return {Promise} resolves to an array wrapping the default assignee's contact id or NULL in case of none.
+     */
+    function getDefaultAssigneeForActivityType (activityType) {
+      switch (activityType.default_assignee_type) {
+        case defaultAssigneeOptionsIndex.SPECIFIC_CONTACT:
+          return $q.resolve([ activityType.default_assignee_contact ]);
+        case defaultAssigneeOptionsIndex.USER_CREATING_THE_CASE:
+          return $q.resolve([ session.contactId ]);
+        case defaultAssigneeOptionsIndex.BY_RELATIONSHIP:
+          return getDefaultAssigneeForActivityTypeByRelationship(activityType);
+        case defaultAssigneeOptionsIndex.NONE:
+        default:
+          return $q.resolve(null);
+      }
+    }
+
+    /**
+     * Returns the default assignee depending on the relationship to the selected
+     * target contact.
+     *
+     * @param {Object} activityType the activity type definition.
+     * @return {Promise}
+     */
+    function getDefaultAssigneeForActivityTypeByRelationship (activityType) {
+      // skip if a target contact has not been selected:
+      if (!$scope.assignment.client_id) {
+        return $q.resolve(null);
+      }
+
+      return Relationship.allValid({
+        'contact_id_a': $scope.assignment.client_id,
+        'relationship_type_id.name_b_a': activityType.default_assignee_relationship,
+        'options': { 'limit': 1 }
+      })
+        .then(function (result) {
+          return result.list.map(function (relationship) {
+            return relationship.contact_id_b;
+          });
+        });
+    }
+
+    /**
+     * Initializes the default assignee for each one of the tasks available.
+     */
+    function initDefaultAssigneesForTasks () {
+      // skip if an activity set and types have not been defined:
+      if (!$scope.activity.activitySet.activityTypes) {
+        return;
+      }
+
+      $scope.activity.activitySet.activityTypes.forEach(function (activityType) {
+        var task = _.find($scope.taskList, function (task) {
+          return task.name === activityType.name;
+        });
+
+        if (!task) {
+          return;
+        }
+
+        getDefaultAssigneeForActivityType(activityType)
+          .then(function (assigneeId) {
+            task.assignee_contact_id = assigneeId;
```

Comment:
Inside a loop, you are doing async operations and assigning values to the loop variable. Are you sure its working perfectly without an IIFE?

Reply:
this is inside a `.forEach` loop, not a `for` loop so it should work.

Reply:
Also as the `task` variable is defined inside the forEach function, I think that is why it works. As its a closure.